crosshair/libimpl/builtinslib.py

```diff
@@ -152,18 +152,39 @@ def smt_coerce(val: Any) -> z3.ExprRef:
 
 
 class SymbolicValue(CrossHairValue):
-    def __init__(self, smtvar: Union[str, z3.ExprRef], typ: Type):
+    def __init__(self, var: z3.ExprRef, typ: Type = None):
+        assert typ is not None
         self.statespace = context_statespace()
         self.snapshot = SnapshotRef(-1)
         self.python_type = typ
-        if type(smtvar) is str:
-            self.var = self.__init_var__(typ, smtvar)
+        self.var = var
+        # TODO test that smtvar's sort matches expected?
+
+    @classmethod
+    def from_z3(cls, var, typ: Type = None):
+        # For more concrete Symbolic objects, we want to
+        # support constructor syntax. To do this we
+        # override __new__, which means we can't just call
+        # cls(var, typ).
+        obj = object.__new__(cls)
+        if typ is None:
+            obj.__init__(var)
         else:
-            self.var = smtvar
-            # TODO test that smtvar's sort matches expected?
+            obj.__init__(var, typ)
```

Comment:
This is the None handling that you were referring to, breaking the csv example, yes?
This `if` statement looks reasonable and necessary to me. You might be thinking that passing `None` into an parameter with a default value causes Python to start using the default value: nope, it will still just receive `None`.

Reply:
The trouble is I'm doing None-handling in both `from_z3` and `__init__`. Ideally I'd keep it in one spot so I don't forget to implement it in one and then not the other.